Firestore/Source/Local/FSTMemoryQueryCache.mm

```diff
@@ -113,11 +118,54 @@ - (nullable FSTQueryData *)queryDataForQuery:(FSTQuery *)query {
   return self.queries[query];
 }
 
+- (void)enumerateSequenceNumbersUsingBlock:(void (^)(FSTListenSequenceNumber sequenceNumber,
+                                                     BOOL *stop))block {
+  [self.queries
+      enumerateKeysAndObjectsUsingBlock:^(FSTQuery *key, FSTQueryData *queryData, BOOL *stop) {
+        block(queryData.sequenceNumber, stop);
+      }];
+  [self.mutatedDocumentSequenceNumbers
+      enumerateKeysAndObjectsUsingBlock:^(FSTDocumentKey *key, NSNumber *sequenceNumber,
+                                          BOOL *stop) {
+        block([sequenceNumber longLongValue], stop);
+      }];
+}
+
+- (NSUInteger)removeQueriesThroughSequenceNumber:(FSTListenSequenceNumber)sequenceNumber
+                                     liveQueries:
+                                         (NSDictionary<NSNumber *, FSTQueryData *> *)liveQueries
+                                           group:(__unused FSTWriteGroup *)group {
+  NSMutableArray<FSTQuery *> *toRemove = [NSMutableArray array];
+  [self.queries
+      enumerateKeysAndObjectsUsingBlock:^(FSTQuery *query, FSTQueryData *queryData, BOOL *stop) {
```

Comment:
I think you can avoid this second scan if your priority were to track (sequence number, target id) pairs. Order by the first (which yields the same sequence numbers) but then once you're done accumulating the queue you have all the primary keys of the targets to delete in hand.

You'll need to store something else to indicate it was a mutated document key instead; not sure if this is required (see other questions).